README.md
[![Build Status](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS_Cyton.svg?branch=master)](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS_Cyton)
[![codecov](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS_Cyton/branch/master/graph/badge.svg)](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS_Cyton)
[![Dependency Status](https://david-dm.org/OpenBCI/OpenBCI_NodeJS_Cyton.svg)](https://david-dm.org/OpenBCI/OpenBCI_NodeJS_Cyton)
[![npm](https://img.shields.io/npm/dm/openbci.svg?maxAge=2592000)](http://npmjs.com/package/@openbci/cyton)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/Flet/semistandard)

# OpenBCI Node.js Cyton SDK

A Node.js module for OpenBCI Cyton ~ written with love by [Push The World!](http://www.pushtheworldllc.com)

We are proud to support all functionality of the Cyton (8 and 16 Channel boards). For the Ganglion (4 channel) please visit [OpenBCI_NodeJS_Ganglion](https://github.com/OpenBCI/OpenBCI_NodeJS_Ganglion). Push The World is actively developing and maintaining this module.

The purpose of this module is to **get connected** and **start streaming** as fast as possible.

Python researcher or developer? Check out how easy it is to [get access to the entire API in the Python example](examples/python)!

### Table of Contents:

---

1. [Installation](#install)
2. [TL;DR](#tldr)
3. [Cyton (32bit Board)](#cyton)
4. [About](#about)
5. [General Overview](#general-overview)
6. [SDK Reference Guide](#sdk-reference-guide)
   - [Constructor](#constructor)
   - [Methods](#method)
   - [Events](#event)
   - [Constants](#constants)
7. [Interfacing With Other Tools](#interfacing-with-other-tools)
8. [Developing](#developing)
9. [Testing](#developing-testing)
10. [Contribute](#contribute)
11. [License](#license)
12. [Roadmap](#roadmap)

### <a name="install"></a> Installation:

```
npm install @openbci/cyton
```

#### serialport dependency

If you encounter this error when trying to run:

```
Error: The module '/path/to/your/project/node_modules/serialport/build/Release/serialport.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 48. This version of Node.js requires
NODE_MODULE_VERSION 51. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or`npm install`).
```

...the issue can be resolved by running:

```
npm rebuild --build-from-source
```

### <a name="tldr"></a> TL;DR:

Get connected and [start streaming right now with the example code](examples/getStreaming/getStreaming.js).

#### Cyton (8 and 16 channel boards)

```ecmascript 6
const Cyton = require('@openbci/cyton');
const ourBoard = new Cyton();
ourBoard.connect(portName) // Port name is a serial port name, see `.listPorts()`
  .then(() => {
    ourBoard.streamStart();
    ourBoard.on('sample',(sample) => {
      /** Work with sample */
      for (let i = 0; i < ourBoard.numberOfChannels(); i++) {
        console.log("Channel " + (i + 1) + ": " + sample.channelData[i].toFixed(8) + " Volts.");
        // prints to the console
        //  "Channel 1: 0.00001987 Volts."
        //  "Channel 2: 0.00002255 Volts."
        //  ...
        //  "Channel 8: -0.00001875 Volts."
      }
    });
});
```

# <a name="cyton"></a> Cyton

## <a name="about"></a> About:

Want to know if the module really works? Check out some projects and organizations using it:

- [_OpenEXP_](https://github.com/openexp/OpenEXP): an open-source desktop app for running experiments and collecting behavioral and physiological data.
- [_Thinker_](http://www.pushtheworldllc.com/): a project building the world's first brainwave-word database.
- [_NeuroJS_](https://github.com/NeuroJS): a community dedicated to Neuroscience research using JavaScript, they have several great examples.

Still not satisfied it works?? Check out this [detailed report](http://s132342840.onlinehome.us/pushtheworld/files/voltageVerificationTestPlanAndResults.pdf) that scientifically validates the output voltages of this module.

How are you still doubting and not using this already? Fine, go look at some of the [800 **_automatic_** tests](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS) written for it!

## <a name="general-overview"></a> General Overview:

Python researcher or developer? Check out how easy it is to [get access to the entire API in the Python example](examples/python)!

## Initialization

Initializing the board:

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
```

Go [checkout out the get streaming example](examples/getStreaming/getStreaming.js)!

For initializing with options, such as verbose print outs:

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton({
  verbose: true
});
```

Or if you don't have a board and want to use synthetic data:

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton({
  simulate: true
});
```

Have a daisy?:

```js
var Cyton = require("@openbci/cyton");
var ourBoard = new Cyton({
  boardType: `daisy`,
  hardSet: true
});
```

Go [checkout out the get streaming with daisy example](examples/getStreamingDaisy/getStreamingDaisy.js)!

Another useful way to start the simulator:

```js
const Cyton = require("@openbci/cyton");
const k = require("@openbci/utilities").constants;
const ourBoard = new Cyton();
ourBoard
  .connect(k.OBCISimulatorPortName) // This will set `simulate` to true
  .then(boardSerial => {
    ourBoard.on("ready", () => {
      /** Start streaming, reading registers, what ever your heart desires  */
    });
  })
  .catch(err => {
    /** Handle connection errors */
  });
```

You can also start the simulator by sending [`.connect(portName)`](#method-connect) with `portName` equal to [`'OpenBCISimulator'`](#constants-obcisimulatorportname).

or if you are using ES6:

```js
import Cyton from "@openbci/cyton";
import { constants } from "@openbci/utilities";
const ourBoard = new Cyton();
ourBoard.connect(constants.OBCISimulatorPortName);
```

To debug, it's amazing, do:

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton({
  debug: true
});
```

Go [checkout out the debug example](examples/debug/debug.js)!

## 'ready' event

You MUST wait for the 'ready' event to be emitted before streaming/talking with the board. The ready happens asynchronously
so installing the 'sample' listener and writing before the ready event might result in... nothing at all.

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
ourBoard
  .connect(portName)
  .then(function(boardSerial) {
    /** Start streaming, reading registers, what ever your heart desires  */
  })
  .catch(function(err) {
    /** Handle connection errors */
  });
```

## Sample properties:

- `startByte` (`Number` should be `0xA0`)
- `sampleNumber` (a `Number` between 0-255)
- `channelData` (channel data indexed at 0 filled with floating point `Numbers` in Volts) if `sendCounts` is false
- `channelDataCounts` (channel data indexed at 0 filled with un scaled interger `Numbers` in raw ADS counts) if `sendCounts` is true
- `accelData` (`Array` with X, Y, Z accelerometer values when new data available) if `sendCounts` is false
- `accelDataCounts` (`Array` with X, Y, Z accelerometer values when new data available) Only present if `sendCounts` is true
- `auxData` (`Buffer` filled with either 2 bytes (if time synced) or 6 bytes (not time synced))
- `stopByte` (`Number` should be `0xCx` where x is 0-15 in hex)
- `boardTime` (`Number` the raw board time)
- `timestamp` (`Number` the `boardTime` plus the NTP calculated offset)

The power of this module is in using the sample emitter, to be provided with samples to do with as you wish.

## To get a ['sample'](#event-sample) event, you need to:

1. Call [`.connect(serialPortName)`](#method-connect)
2. Install the ['ready'](#event-ready) event emitter on resolved promise
3. In callback for ['ready'](#event-ready) emitter, call [`streamStart()`](#method-stream-start)
4. Install the ['sample'](#event-sample) event emitter

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
ourBoard
  .connect(portName)
  .then(function() {
    ourBoard.streamStart();
    ourBoard.on("sample", function(sample) {
      /** Work with sample */
    });
  })
  .catch(function(err) {
    /** Handle connection errors */
  });
```

Close the connection with [`.streamStop()`](#method-stream-stop) and disconnect with [`.disconnect()`](#method-disconnect)

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
ourBoard.streamStop().then(ourBoard.disconnect());
```

## Time Syncing

You must be using OpenBCI firmware version 2 in order to do time syncing. After you [`.connect()`](#method-connect) and send a [`.softReset()`](#method-soft-reset), you can call [`.usingVersionTwoFirmware()`](#method-using-version-two-firmware) to get a boolean response as to if you are using `v1` or `v2`.

Now using firmware `v2`, the fun begins! We synchronize the Board's clock with the module's time. In firmware `v2` we leverage samples with time stamps and _ACKs_ from the Dongle to form a time synchronization strategy. Time syncing has been verified to +/- 4ms and a test report is on the way. We are still working on the synchronize of this module and an NTP server, this is an open call for any NTP experts out there! With a global NTP server you could use several different devices and all sync to the same time server. That way you can really do some serious cloud computing!

Keep your resync interval above 50ms. While it's important to resync every couple minutes due to drifting of clocks, please do not try to sync without getting the last sync event! We can only support one sync operation at a time!

Using local computer time:

```js
const Cyton = require("@openbci/cyton");
const k = require("@openbci/utilities").constants;
const ourBoard = new Cyton({
  verbose: true
});

const resyncPeriodMin = 5; // re sync every five minutes
const secondsInMinute = 60;
let sampleRate = k.OBCISampleRate250; // Default to 250, ALWAYS verify with a call to `.sampleRate()` after 'ready' event!
let timeSyncPossible = false;

// Call to connect
ourBoard
  .connect(portName)
  .then(() => {
    // Get the sample rate after 'ready'
    sampleRate = ourBoard.sampleRate();
    // Find out if you can even time sync, you must be using v2 and this is only accurate after a `.softReset()` call which is called internally on `.connect()`. We parse the `.softReset()` response for the presence of firmware version 2 properties.
    timeSyncPossible = ourBoard.usingVersionTwoFirmware();

    ourBoard
      .streamStart()
      .then(() => {
        /** Start streaming command sent to board. */
      })
      .catch(err => {
        console.log(`stream start: ${err}`);
      });

    // PTW recommends sample driven
    ourBoard.on("sample", sample => {
      // Resynchronize every every 5 minutes
      if (
        sample._count %
          (sampleRate * resyncPeriodMin * secondsInMinute) ===
        0
      ) {
        ourBoard.syncClocksFull().then(syncObj => {
          // Sync was successful
          if (syncObj.valid) {
            // Log the object to check it out!
            console.log(`syncObj`, syncObj);

            // Sync was not successful
          } else {
            // Retry it
            console.log(`Was not able to sync, please retry?`);
          }
        });
      }

      if (sample.timeStamp) {
        // true after the first sync
        console.log(`NTP Time Stamp ${sample.timeStamp}`);
      }
    });
  })
  .catch(err => {
    console.log(`connect: ${err}`);
  });
```

## Auto-finding boards

You must have the OpenBCI board connected to the PC before trying to automatically find it.

If a port is not automatically found, then call [`.listPorts()`](#method-list-ports) to get a list of all serial ports this would be a good place to present a drop down picker list to the user, so they may manually select the serial port name.

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
ourBoard.autoFindOpenBCIBoard().then(portName => {
  if (portName) {
    /**
     * Connect to the board with portName
     * i.e. ourBoard.connect(portName).....
     */
  } else {
    /**Unable to auto find OpenBCI board*/
  }
});
```

Note: [`.autoFindOpenBCIBoard()`](#method-auto-find-open-bci-board) will return the name of the Simulator if you instantiate with option `simulate: true`.

## Auto Test - (Using impedance to determine signal quality)

Measuring impedance is a vital tool in ensuring great data is collected.

**_IMPORTANT!_** Measuring impedance takes time, so _only test what you must_

Your OpenBCI board will have electrodes hooked up to either a P input, N input or in some cases both inputs.

To test specific inputs of channels:

1. Connect to board.
2. Start streaming.
3. Install the ['impedanceArray'](#event-impedance-array) event
4. Call [`.impedanceTestChannels()`](#method-impedance-test-channels) with your configuration array

A configuration array looks like, for an 8 channel board, `['-','N','n','p','P','-','b','b']`

Where there are the same number of elements as channels and each element can be either:

- `p` or `P` (only test P input)
- `n` or `N` (only test N input)
- `b` or `B` (test both inputs) (takes 66% longer to run then previous two `p` or `n`)
- `-` (ignore channel)

Without further ado, here is an example:

```js
const Cyton = require("@openbci/cyton");
const ourBoard = new Cyton();
ourBoard
  .connect(portName)
  .then(function(boardSerial) {
    ourBoard.streamStart();
    ourBoard.once("impedanceArray", impedanceArray => {
      /** Work with impedance Array */
    });
    ourBoard
      .impedanceTestChannels(["n", "N", "n", "p", "P", "p", "b", "B"])
      .catch(err => console.log(err));
  })
  .catch(function(err) {
    /** Handle connection errors */
  });
```

But wait! What is this `impedanceArray`? An Array of Objects, for each object:

```
[{
    channel: 1,
    P: {
        raw: -1,
        text: 'init'
    },
    N: {
        raw: -1,
        text: 'init'
    }
},
{
    // Continues for each channel up to the amount of channels on board (8 or 16)
},...];
```

Where:

- _channel_ is the channel number (`impedanceArray[0]` is channel 1, `impedanceArray[6]` is channel 7)
- _P_ is the P input data (Note: P is capitalized)
  - _raw_ is an impedance value resulting from the Goertzel algorithm.
  - _text_ is a text interpretation of the `average`
    - **Good** impedance is < 5k Ohms
    - **Ok** impedance is 5 to 10k Ohms
    - **Bad** impedance is > 10k Ohms
    - **None** impedance is > 1M Ohms
- _N_ is the N input data (Note: N is capitalized) (see above for what N object consists of)

To run an impedance test on all inputs, one channel at a time:

1. Connect to board
2. Start streaming
3. Install the ['impedanceArray'](#event-impedance-array)
4. Call [`.impedanceTestAllChannels()`](#method-impedance-test-all-channels)

**Note: Takes up to 5 seconds to start measuring impedances. There is an unknown number of samples taken. Not always 60!**

For example:

```js
const Cyton = require('@openbci/cyton');
const ourBoard = new Cyton();
ourBoard.connect(portName).then(function(boardSerial) {
    ourBoard.streamStart();
    ourBoard.on('impedanceArray', impedanceArray => {
        /** Work with impedance */
    });
    ourBoard.impedanceTestAllChannels();
}
```

See Reference Guide for a complete list of impedance tests.

## <a name="sdk-reference-guide"></a> SDK Reference Guide:

---

### <a name="constructor"></a> Constructor:

#### <a name="init"></a> Cyton (options)

Create new instance of an Cyton board.

**_options (optional)_**

Board optional configurations.

- `baudRate` {Number} - Baud Rate, defaults to 115200. Manipulating this is allowed if firmware on board has been previously configured.
- `boardType` {String} - Specifies type of OpenBCI board (3 possible boards)
  - `default` - 8 Channel OpenBCI board (Default)
  - `daisy` - 8 Channel board with Daisy Module - 16 Channels
- `hardSet` {Boolean} - Recommended if using `daisy` board! For some reason, the `daisy` is sometimes not picked up by the module so you can set `hardSet` to true which will ensure the daisy is picked up. (Default `false`)
- `simulate` {Boolean} - Full functionality, just mock data. Must attach Daisy module by setting `simulatorDaisyModuleAttached` to `true` in order to get 16 channels. (Default `false`)
- `simulatorBoardFailure` {Boolean} - Simulates board communications failure. This occurs when the RFduino on the board is not polling the RFduino on the dongle. (Default `false`)
- `simulatorDaisyModuleAttached` {Boolean} - Simulates a daisy module being attached to the OpenBCI board. This is useful if you want to test how your application reacts to a user requesting 16 channels but there is no daisy module actually attached, or vice versa, where there is a daisy module attached and the user only wants to use 8 channels. (Default `false`)
- `simulatorFirmwareVersion` {String} - Allows the simulator to use firmware version 2 features. (2 Possible Options)
  - `v1` - Firmware Version 1 (Default)
  - `v2` - Firmware Version 2
- `simulatorFragmentation` {String} - Specifies how to break packets to simulate fragmentation, which occurs commonly in real devices. It is recommended to test code with this enabled. (4 Possible Options)
  - `none` - do not fragment packets; output complete chunks immediately when produced (Default)
  - `random` - output random small chunks of data interspersed with full buffers
  - `fullBuffers` - allow buffers to fill up until latency timer has expired
  - `oneByOne` - output each byte separately
- `simulatorLatencyTime` {Number} - The time in milliseconds to wait before sending partially full buffers of data, if `simulatorFragmentation` is specified. (Default `16`)
- `simulatorBufferSize` {Number} - The size of a full buffer of data, if `simulatorFragmentation` is specified. (Default `4096`)

changelog.md
# 2.0.1

### Bug Fixes

- Fixing references to openbci-utilities and openbci-cyton

# 2.0.0

### Breaking Changes

- Name of package is now @openbci/cyton

### Enhancements

- Include node 10 in testing

# 1.1.3

### Bug Fixes

- Connect function failed and forced close of port when serial port flushed too fast

# 1.1.2

### Bug Fixes

- Connected function failed

# 1.1.1

Bumping the mathjs to 4.0.1 to fix insecurity

# 1.1.0

Bumping the minor version of openbci-utilities allows this module to move forward

### Breaking Changes

- Drop support for node 4 because of new OpenBCI utilities
- Connect will now automatically call `softReset` and resolve once it's complete.

# 1.0.8

Bump serial port to 1.0.8

### Bug Fixes

- Fix `python` example

# 1.0.7

### Bug Fixes

- Add simulator back into this repo
- Fixed bug where timestamp was not calculated with timeOffset, big fail.
- Fixed the `python` example
- Fixed the `labstreaminglayer` example

# 1.0.6

### Bug Fixes

- Daisy samples now get stop bytes with utilities bump to v0.2.7

# 1.0.4/5

### New Features

- Add function `impedanceSet` which is similar to `channelSet` in that it just sends commands and then forgets about them.

### Bug Fixes

- Fixed lead off command sending for v2 firmware and later.

# 1.0.3

### Bug Fixes

- Bumped utilities module to v0.2.1
- Fixes issue with `getStreamingDaisy.js`

# 1.0.2

### Chores

- Removed false badges

# 1.0.1

### Bug Fixes

- Problem in package.json prevented publishing

# 1.0.0

Port `openbci` v2.2.0 to this repo. Please see the change log for `openbci` if you need a history of changes.


openBCISimulator.js
'use strict';
const EventEmitter = require('events').EventEmitter;
const util = require('util');
const stream = require('stream');

const obciUtilities = require('@openbci/utilities/dist/utilities');
const k = require('@openbci/utilities/dist/constants');
const now = require('performance-now');
const Buffer = require('safe-buffer').Buffer;

const _options = {
  accel: true,
  alpha: true,
  boardFailure: false,
  daisy: false,
  daisyCanBeAttached: true,
  drift: 0,
  firmwareVersion: [k.OBCIFirmwareV1, k.OBCIFirmwareV2, k.OBCIFirmwareV3],
  fragmentation: [k.OBCISimulatorFragmentationNone, k.OBCISimulatorFragmentationRandom, k.OBCISimulatorFragmentationFullBuffers, k.OBCISimulatorFragmentationOneByOne],
  latencyTime: 16,
  bufferSize: 4096,
  lineNoise: [k.OBCISimulatorLineNoiseHz60, k.OBCISimulatorLineNoiseHz50, k.OBCISimulatorLineNoiseNone],
  sampleRate: 250,
  serialPortFailure: false,
  verbose: false
};

function Simulator (portName, options) {
  if (!(this instanceof Simulator)) {
    return new Simulator(portName, options);
  }
  options = options || {};
  let opts = {};

  stream.Stream.call(this);

  /** Configuring Options */
  let o;
  for (o in _options) {
    let userValue = options[o];
    delete options[o];

    if (typeof _options[o] === 'object') {
      // an array specifying a list of choices
      // if the choice is not in the list, the first one is defaulted to

      if (_options[o].indexOf(userValue) !== -1) {
        opts[o] = userValue;
      } else {
        opts[o] = _options[o][0];
      }
    } else {
      // anything else takes the user value if provided, otherwise is a default

      if (userValue !== undefined) {
        opts[o] = userValue;
      } else {
        opts[o] = _options[o];
      }
    }
  }

  for (o in options) throw new Error('"' + o + '" is not a valid option');

  this.options = opts;

  // Bools
  this.isOpen = false;
  this.sd = {
    active: false,
    startTime: 0
  };
  this.streaming = false;
  this.synced = false;
  this.sendSyncSetPacket = false;
  // Buffers
  this.outputBuffer = new Buffer(this.options.bufferSize);
  this.outputBuffered = 0;
  // Numbers
  this.channelNumber = 1;
  this.hostChannelNumber = this.channelNumber;
  this.pollTime = 80;
  this.sampleNumber = -1; // So the first sample is 0
  // Objects
  this.sampleGenerator = obciUtilities.randomSample(k.OBCINumberOfChannelsDefault, this.options.sampleRate, this.options.alpha, this.options.lineNoise);
  this.time = {
    current: 0,
    start: now(),
    loop: null
  };
  // Strings
  this.portName = portName || k.OBCISimulatorPortName;

  // Call 'open'
  if (this.options.verbose) console.log(`Port name: ${portName}`);
  setTimeout(() => {
    this.isOpen = true;
    this.emit('open');
  }, 200);
}

// This allows us to use the emitter class freely outside of the module
util.inherits(Simulator, EventEmitter);

Simulator.prototype.flush = function (callback) {
  this.outputBuffered = 0;

  clearTimeout(this.outputLoopHandle);
  this.outputLoopHandle = null;

  if (callback) callback();
};

// output only size bytes of the output buffer
Simulator.prototype._partialDrain = function (size) {
  if (!this.isOpen) throw new Error('not connected');

  if (size > this.outputBuffered) size = this.outputBuffered;

  // buffer is copied because presently openBCICyton.js reuses it
  let outBuffer = new Buffer(this.outputBuffer.slice(0, size));

  this.outputBuffer.copy(this.outputBuffer, 0, size, this.outputBuffered);
  this.outputBuffered -= size;

  this.emit('data', outBuffer);
};

// queue some data for output and send it out depending on options.fragmentation
Simulator.prototype._output = function (dataBuffer) {
  // drain full buffers until there is no overflow
  while (this.outputBuffered + dataBuffer.length > this.outputBuffer.length) {
    let len = dataBuffer.copy(this.outputBuffer, this.outputBuffered);
    dataBuffer = dataBuffer.slice(len);
    this.outputBuffered += len;

    this._partialDrain(this.outputBuffered);
    this.flush();
  }

  dataBuffer.copy(this.outputBuffer, this.outputBuffered);
  this.outputBuffered += dataBuffer.length;

  if (!this.outputLoopHandle) {
    let latencyTime = this.options.latencyTime;
    if (this.options.fragmentation === k.OBCISimulatorFragmentationOneByOne ||
      this.options.fragmentation === k.OBCISimulatorFragmentationNone) {
      // no need to wait for latencyTime
      // note that this is the only difference between 'none' and 'fullBuffers'
      latencyTime = 0;
    }
    let outputLoop = () => {
      let size;
      switch (this.options.fragmentation) {
        case k.OBCISimulatorFragmentationRandom:
          if (Math.random() < 0.5) {
            // randomly picked to send out a fragment
            size = Math.ceil(Math.random() * Math.max(this.outputBuffered, 62));
            break;
          } // else, randomly picked to send a complete buffer in next block
        /* falls through */
        case k.OBCISimulatorFragmentationFullBuffers:
        case k.OBCISimulatorFragmentationNone:
        case false:
          size = this.outputBuffered;
          break;
        case k.OBCISimulatorFragmentationOneByOne:
          size = 1;
          break;
      }
      this._partialDrain(size);
      if (this.outputBuffered) {
        this.outputLoopHandle = setTimeout(outputLoop, latencyTime);
      } else {
        this.outputLoopHandle = null;
      }
    };
    if (latencyTime === 0) {
      outputLoop();
    } else {
      this.outputLoopHandle = setTimeout(outputLoop, latencyTime);
    }
  }
};

Simulator.prototype.write = function (data, callback) {
  if (!this.isOpen) {
    /* istanbul ignore else */
    if (callback) callback(Error('Not connected'));
    else throw new Error('Not connected!');
    return;
  }

  // TODO: this function assumes a type of Buffer for radio, and a type of String otherwise
  //       FIX THIS it makes it unusable outside the api code
  switch (data[0]) {
    case k.OBCIRadioKey:
      this._processPrivateRadioMessage(data);
      break;
    case k.OBCIStreamStart:
      if (!this.stream) this._startStream();
      this.streaming = true;
      break;
    case k.OBCIStreamStop:
      if (this.stream) clearInterval(this.stream); // Stops the stream
      this.streaming = false;
      break;
    case k.OBCIMiscSoftReset:
      if (this.stream) clearInterval(this.stream);
      this.streaming = false;
      this._output(new Buffer(`OpenBCI V3 Simulator On Board ADS1299 Device ID: 0x3E ${this.options.daisy ? `On Daisy ADS1299 Device ID: 0x3E\n` : ``} LIS3DH Device ID: 0x38422 ${this.options.firmwareVersion === k.OBCIFirmwareV2 ? `Firmware: v2.0.0\n` : ``}$$$`));
      break;
    case k.OBCISDLogForHour1:
    case k.OBCISDLogForHour2:
    case k.OBCISDLogForHour4:
    case k.OBCISDLogForHour12:
    case k.OBCISDLogForHour24:
    case k.OBCISDLogForMin5:
    case k.OBCISDLogForMin15:
    case k.OBCISDLogForMin30:
    case k.OBCISDLogForSec14:
      // If we are not streaming, then do verbose output
      if (!this.streaming) {
        this._output(new Buffer('Wiring is correct and a card is present.\nCorresponding SD file OBCI_69.TXT\n$$$'));
      }
      this.sd.active = true;
      this.sd.startTime = now();
      break;
    case k.OBCISDLogStop:
      if (!this.streaming) {
        if (this.SDLogActive) {
          this._output(new Buffer(`Total Elapsed Time: ${now() - this.sd.startTime} ms`));
          this._output(new Buffer(`Max write time: ${Math.random() * 500} us`));
          this._output(new Buffer(`Min write time: ${Math.random() * 200} us`));
          this._output(new Buffer(`Overruns: 0`));
          this._printEOT();
        } else {
          this._output(new Buffer('No open file to close\n'));
          this._printEOT();
        }
      }
      this.SDLogActive = false;
      break;
    case k.OBCISyncTimeSet:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        this.synced = true;
        setTimeout(() => {
          this._output(new Buffer(k.OBCISyncTimeSent));
          this._syncUp();
        }, 10);
      }
      break;
    case k.OBCIChannelMaxNumber8:
      if (this.options.daisy) {
        this.options.daisy = false;
        this._output(new Buffer(k.OBCIChannelMaxNumber8SuccessDaisyRemoved));
        this._printEOT();
      } else {
        this._printEOT();
      }
      break;
    case k.OBCIChannelMaxNumber16:
      if (this.options.daisy) {
        this._output(new Buffer(k.OBCIChannelMaxNumber16DaisyAlreadyAttached));
        this._printEOT();
      } else {
        if (this.options.daisyCanBeAttached) {
          this.options.daisy = true;
          this._output(new Buffer(k.OBCIChannelMaxNumber16DaisyAttached));
          this._printEOT();
        } else {
          this._output(new Buffer(k.OBCIChannelMaxNumber16NoDaisyAttached));
          this._printEOT();
        }
      }
      break;
    case k.OBCIMiscQueryRegisterSettings:
      let outputString = k.OBCIRegisterQueryCyton;
      if (this.options.daisy) {
        outputString += k.OBCIRegisterQueryCytonDaisy;
      }
      if (this.options.firmwareVersion === k.OBCIFirmwareV3) {
        outputString += k.OBCIRegisterQueryAccelerometerFirmwareV3;
      } else {
        outputString += k.OBCIRegisterQueryAccelerometerFirmwareV1;
      }
      this._output(Buffer.from(outputString));
      this._printEOT();
      break;
    default:
      break;
  }

  /** Handle Callback */
  if (callback) {
    callback(null, 'Success!');
  }
};

Simulator.prototype.drain = function (callback) {
  if (callback) callback();
};

Simulator.prototype.close = function (callback) {
  if (this.isOpen) {
    this.flush();

    if (this.stream) clearInterval(this.stream);

    this.isOpen = false;
    this.emit('close');
    if (callback) callback();
  } else {
    if (callback) callback(Error('Not connected'));
  }
};

Simulator.prototype._startStream = function () {
  let intervalInMS = 1000 / this.options.sampleRate;

  if (intervalInMS < 2) intervalInMS = 2;

  let getNewPacket = sampNumber => {
    if (this.options.accel) {
      if (this.synced) {
        if (this.sendSyncSetPacket) {
          this.sendSyncSetPacket = false;
          return obciUtilities.convertSampleToPacketAccelTimeSyncSet(this.sampleGenerator(sampNumber), now().toFixed(0));
        } else {
          return obciUtilities.convertSampleToPacketAccelTimeSynced(this.sampleGenerator(sampNumber), now().toFixed(0));
        }
      } else {
        return obciUtilities.convertSampleToPacketStandard(this.sampleGenerator(sampNumber));
      }
    } else {
      if (this.synced) {
        if (this.sendSyncSetPacket) {
          this.sendSyncSetPacket = false;
          return obciUtilities.convertSampleToPacketRawAuxTimeSyncSet(this.sampleGenerator(sampNumber), now().toFixed(0), new Buffer([0, 0, 0, 0, 0, 0]));
        } else {
          return obciUtilities.convertSampleToPacketRawAuxTimeSynced(this.sampleGenerator(sampNumber), now().toFixed(0), new Buffer([0, 0, 0, 0, 0, 0]));
        }
      } else {
        return obciUtilities.convertSampleToPacketRawAux(this.sampleGenerator(sampNumber), new Buffer([0, 0, 0, 0, 0, 0]));
      }
    }
  };

  this.stream = setInterval(() => {
    this._output(getNewPacket(this.sampleNumber));
    this.sampleNumber++;
  }, intervalInMS);
};

Simulator.prototype._syncUp = function () {
  setTimeout(() => {
    this.sendSyncSetPacket = true;
  }, 12); // 3 packets later
};

Simulator.prototype._printEOT = function () {
  this._output(new Buffer('$$$'));
};

Simulator.prototype._printFailure = function () {
  this._output(new Buffer('Failure: '));
};

Simulator.prototype._printSuccess = function () {
  this._output(new Buffer('Success: '));
};

Simulator.prototype._printValidatedCommsTimeout = function () {
  this._printFailure();
  this._output(new Buffer('Communications timeout - Device failed to poll Host'));
  this._printEOT();
};

Simulator.prototype._processPrivateRadioMessage = function (dataBuffer) {
  switch (dataBuffer[1]) {
    case k.OBCIRadioCmdChannelGet:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (!this.options.boardFailure) {
          this._printSuccess();
          this._output(new Buffer(`Host and Device on Channel Number ${this.channelNumber}`));
          this._output(new Buffer([this.channelNumber]));
          this._printEOT();
        } else if (!this.serialPortFailure) {
          this._printFailure();
          this._output(new Buffer(`Host on Channel Number ${this.channelNumber}`));
          this._output(new Buffer([this.channelNumber]));
          this._printEOT();
        }
      }
      break;
    case k.OBCIRadioCmdChannelSet:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (!this.options.boardFailure) {
          if (dataBuffer[2] <= k.OBCIRadioChannelMax) {
            this.channelNumber = dataBuffer[2];
            this.hostChannelNumber = this.channelNumber;
            this._printSuccess();
            this._output(new Buffer(`Channel Number ${this.channelNumber}`));
            this._output(new Buffer([this.channelNumber]));
            this._printEOT();
          } else {
            this._printFailure();
            this._output(new Buffer('Verify channel number is less than 25'));
            this._printEOT();
          }
        } else if (!this.serialPortFailure) {
          this._printValidatedCommsTimeout();
        }
      }
      break;
    case k.OBCIRadioCmdChannelSetOverride:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (dataBuffer[2] <= k.OBCIRadioChannelMax) {
          if (dataBuffer[2] === this.channelNumber) {
            this.options.boardFailure = false;
          } else {
            this.options.boardFailure = true;
          }
          this.hostChannelNumber = dataBuffer[2];
          this._printSuccess();
          this._output(new Buffer(`Host override - Channel Number ${this.hostChannelNumber}`));
          this._output(new Buffer([this.hostChannelNumber]));
          this._printEOT();
        } else {
          this._printFailure();
          this._output(new Buffer('Verify channel number is less than 25'));
          this._printEOT();
        }
      }
      break;
    case k.OBCIRadioCmdPollTimeGet:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (!this.options.boardFailure) {
          this._printSuccess();
          this._output(new Buffer(`Poll Time ${this.pollTime}`));
          this._output(new Buffer([this.pollTime]));
          this._printEOT();
        } else {
          this._printValidatedCommsTimeout();
        }
      }
      break;
    case k.OBCIRadioCmdPollTimeSet:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (!this.options.boardFailure) {
          this.pollTime = dataBuffer[2];
          this._printSuccess();
          this._output(new Buffer(`Poll Time ${this.pollTime}`));
          this._output(new Buffer([this.pollTime]));
          this._printEOT();
        } else {
          this._printValidatedCommsTimeout();
        }
      }
      break;
    case k.OBCIRadioCmdBaudRateSetDefault:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        this._printSuccess();
        this._output(new Buffer('Switch your baud rate to 115200'));
        this._output(new Buffer([0x24, 0x24, 0x24, 0xFF])); // The board really does this
      }
      break;
    case k.OBCIRadioCmdBaudRateSetFast:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        this._printSuccess();
        this._output(new Buffer('Switch your baud rate to 230400'));
        this._output(new Buffer([0x24, 0x24, 0x24, 0xFF])); // The board really does this
      }
      break;
    case k.OBCIRadioCmdSystemStatus:
      if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
        if (!this.options.boardFailure) {
          this._printSuccess();
          this._output(new Buffer('System is Up'));
          this._printEOT();
        } else {
          this._printFailure();
          this._output(new Buffer('System is Down'));
          this._printEOT();
        }
      }
      break;
    default:
      break;
  }
};

module.exports = Simulator;
